File: src/frontend/src/api/runtimeProviders.ts
import type { UUID } from './types'

const API_BASE = (import.meta.env.VITE_API_BASE_URL ?? '').replace(/\/$/, '')

export interface RuntimeProviderStatus {
  ready: boolean
  code: string
  message: string
  installable: boolean
}

export interface TextRuntimeProvider {
  mode: 'local' | 'cloud'
  base_url: string
  model: string
  context_window: number
  has_api_key: boolean
  status: RuntimeProviderStatus
}

export interface ImageRuntimeProvider {
  mode: 'local' | 'cloud' | 'off'
  base_url: string
  model: string
  has_api_key: boolean
  status: RuntimeProviderStatus
}

export interface RuntimeProviderProfile {
  text: TextRuntimeProvider
  image: ImageRuntimeProvider
}

export interface InstallJob {
  id: string
  kind: 'text' | 'image'
  status: 'running' | 'completed' | 'failed'
  error?: string | null
  result?: RuntimeProviderStatus | null
  created_at: string
}

async function request<T>(path: string, init?: RequestInit): Promise<T> {
  const response = await fetch(`${API_BASE}${path}`, {
    ...init,
    headers: {
      'Content-Type': 'application/json',
      ...(init?.headers ?? {}),
    },
  })
  if (!response.ok) {
    let message = `HTTP ${response.status}`
    try {
      const payload = await response.json() as { detail?: unknown }
      if (typeof payload.detail === 'string') message = payload.detail
    } catch { /* keep generic HTTP message */ }
    throw new Error(message)
  }
  return response.json() as Promise<T>
}

export const runtimeProviderApi = {
  profile: () => request<RuntimeProviderProfile>('/api/runtime/providers'),
  check: () => request<{ text: RuntimeProviderStatus; image: RuntimeProviderStatus }>(
    '/api/runtime/providers/check',
    { method: 'POST' },
  ),
  configureText: (payload: {
    mode: 'local' | 'cloud'
    base_url?: string
    model?: string
    api_key?: string
    context_window?: number
    campaign_id?: UUID
  }) => request<TextRuntimeProvider>('/api/runtime/providers/text', {
    method: 'PUT',
    body: JSON.stringify(payload),
  }),
  configureImage: (payload: {
    mode: 'local' | 'cloud' | 'off'
    base_url?: string
    model?: string
    api_key?: string
  }) => request<ImageRuntimeProvider>('/api/runtime/providers/image', {
    method: 'PUT',
    body: JSON.stringify(payload),
  }),
  install: (kind: 'text' | 'image') => request<InstallJob>(
    `/api/runtime/providers/${kind}/install`,
    { method: 'POST' },
  ),
  installJob: (jobId: string) => request<InstallJob>(`/api/runtime/providers/install/${jobId}`),
}

File: src/backend/app/api/runtime_providers.py
from __future__ import annotations

import asyncio
import uuid
from dataclasses import dataclass, field
from datetime import datetime
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, Field
from sqlalchemy.ext.asyncio import AsyncSession

from app.db.engine import get_session
from app.models.provider_config import ProviderConfigCreate
from app.services.campaign_service import CampaignService
from app.services.runtime_provider_service import RuntimeProviderError, RuntimeProviderService

router = APIRouter(prefix="/api/runtime/providers", tags=["runtime-providers"])


class TextProviderUpdate(BaseModel):
    mode: str = Field(pattern="^(local|cloud)$")
    base_url: str | None = None
    model: str | None = None
    api_key: str | None = None
    context_window: int | None = Field(default=None, ge=1024)
    campaign_id: UUID | None = None


class ImageProviderUpdate(BaseModel):
    mode: str = Field(pattern="^(local|cloud|off)$")
    base_url: str | None = None
    model: str | None = None
    api_key: str | None = None


@dataclass
class InstallJob:
    id: str
    kind: str
    status: str = "running"
    error: str | None = None
    result: dict | None = None
    created_at: str = field(default_factory=lambda: datetime.utcnow().isoformat())


_jobs: dict[str, InstallJob] = {}


def _job_payload(job: InstallJob) -> dict:
    return {
        "id": job.id,
        "kind": job.kind,
        "status": job.status,
        "error": job.error,
        "result": job.result,
        "created_at": job.created_at,
    }


async def _run_install(job: InstallJob) -> None:
    service = RuntimeProviderService()
    try:
        if job.kind == "text":
            result = await asyncio.to_thread(service.ensure_local_text)
        else:
            result = await asyncio.to_thread(service.ensure_local_image)
        job.result = result
        job.status = "completed"
    except Exception as exc:  # install boundary: return actionable failure to UI
        job.error = str(exc)
        job.status = "failed"


@router.get("")
async def get_runtime_providers():
    return await asyncio.to_thread(RuntimeProviderService().profile)


@router.post("/check")
async def check_runtime_providers():
    service = RuntimeProviderService()
    return {
        "text": await asyncio.to_thread(service.check_text),
        "image": await asyncio.to_thread(service.check_image),
    }


@router.put("/text")
async def update_text_provider(
    data: TextProviderUpdate,
    session: AsyncSession = Depends(get_session),
):
    service = RuntimeProviderService()
    try:
        profile = service.configure_text(
            data.mode,
            base_url=data.base_url,
            model=data.model,
            api_key=data.api_key,
            context_window=data.context_window,
        )
        if data.campaign_id:
            await CampaignService(session).configure_provider(
                data.campaign_id,
                ProviderConfigCreate(
                    base_url=profile["base_url"],
                    model_name=profile["model"],
                    api_key=(
                        data.api_key
                        if data.mode == "cloud"
                        else None
                    ),
                    context_window=profile["context_window"],
                ),
            )
        return service.profile()["text"]
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


@router.put("/image")
async def update_image_provider(data: ImageProviderUpdate):
    service = RuntimeProviderService()
    try:
        service.configure_image(
            data.mode,
            base_url=data.base_url,
            model=data.model,
            api_key=data.api_key,
        )
        return service.profile()["image"]
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


@router.post("/{kind}/install")
async def install_local_provider(kind: str):
    if kind not in {"text", "image"}:
        raise HTTPException(status_code=404, detail="Unknown provider kind")
    service = RuntimeProviderService()
    mode = service.text_mode() if kind == "text" else service.image_mode()
    if mode != "local":
        raise HTTPException(status_code=409, detail=f"{kind} provider is not configured as local")
    job = InstallJob(id=str(uuid.uuid4()), kind=kind)
    _jobs[job.id] = job
    asyncio.create_task(_run_install(job), name=f"provider-install-{kind}-{job.id}")
    return _job_payload(job)


@router.get("/install/{job_id}")
async def get_install_job(job_id: str):
    job = _jobs.get(job_id)
    if not job:
        raise HTTPException(status_code=404, detail="Install job not found")
    return _job_payload(job)
